### PR TITLE
Use merge-base in find_commit.dart

### LINKED ISSUE #97595: customer_testing doesn't update for merges
The `customer_testing` test uses a version of flutter/test determined by the initial branch point of the PR, found via https://github.com/flutter/flutter/blob/master/dev/tools/bin/find_commit.dart:
https://github.com/flutter/flutter/blob/72e426fc7862db258d3c42b0701e7fbfdf671575/dev/customer_testing/ci.sh#L21-L34

This causes problems in PRs for people who use a merge workflow to get new versions of `master` (which is the only one supported by the GitHub UI); the merge will make the PR up to date for the purposes of every test except for `customer_testing`, creating exactly the kind of mismatch that the script is intended to avoid. The only way to fix a PR in that state is to rebase (from a local checkout) and force-push, which is much less convenient. It's also a very weird failure state; the only way to know this is to have it happen to you, be confused, and ask, which is a frustrating experience for developers.

If possible, when there has been a merge from the original branch `find_commit` should use the last merged commit instead of the original branch commit to determine the anchor date.

### PR DESCRIPTION
**Pull request by natebosch**
Closes #97595

The prior approach of manually diffing the entire log chain is less
efficient, and only found the original branch point ignoring subsequent
merges. The limitation forced PR workflows into rebasing and force
pushing new history to get the branch point far enough for CI to pass.

Use `git merge-base` to find the latest common commit with the main
branch.
Add an `allowFailure` argument to the `git` utility to use a more
specific failure in the case of no shared history when this command will
fail with a generic error.

Use `^branch` with the `git log` commands to exclude shared history and
more easily count the unique commits on each branch.

Drop the `Commit` abstraction. Parse directly to timestamp or line counts.

### CODE REVIEW

File: dev/tools/bin/find_commit.dart [hunk 1]
```diff
@@ -43,41 +29,37 @@ String findCommit({
...
-    if (branchCommits.isEmpty || trunkCommits.isEmpty || branchCommits.first.hash != trunkCommits.first.hash) {
+    final String mergeBase = git(primaryRepoDirectory, <String>['merge-base', primaryBranch, primaryTrunk], allowFailure: true).trim();
+    if (mergeBase.isEmpty) {
       throw StateError('Branch $primaryBranch does not seem to have a common history with trunk $primaryTrunk.');
     }
-    if (branchCommits.last.hash == trunkCommits.last.hash) {
-      log('$primaryBranch is even with $primaryTrunk, using last commit as anchor');
-      anchor = trunkCommits.last;
+    anchor = _parseTimestamp(git(primaryRepoDirectory, <String>['log', _commitTimestampFormat, '--max-count=1', mergeBase, '--']));
+    final int missingTrunkCommits = _countLines(git(primaryRepoDirectory, <String>['rev-list', primaryTrunk, '^$primaryBranch', '--']));
```
Comment (natebosch) [inline dev/tools/bin/find_commit.dart:42]: we could skip these processes by surrounding this with a `debugLogging` check now, I may as well add that optimization.

**Review by stuartmorgan-g — COMMENTED** [review]
Looks reasonable to me, but @Hixie should review since he's more familiar with the script and the needs of `customer_test`.

**Comment by Hixie** [comment]
The git parts of this are all arcane magic to me, unfortunately.

How did you test this? Does it correctly pick the right branch in common cases? What cases would it fail in? Are there cases where this will find a commit that isn't the one that indicates what the branch is up to date to?

**Review by Hixie — APPROVED** [review]
RSLGTM
...modulo answers to the questions asked above

**Comment by natebosch** [comment]
> How did you test this?

Manual testing locally with some repos set up in various states of orphaned or branched and merged commits.



> Does it correctly pick the right branch in common cases?

Yes, it seems to pick the correct commit in common cases, and a better commit in merge cases.



> What cases would it fail in?

I don't think I can imagine any, AFAICT `merge-base` does exactly what we'd want based on my understanding of the intent for this tool, but without existing tests I don't know if there is some edge I'm overlooking.



> Are there cases where this will find a commit that isn't the one that indicates what the branch is up to date to?

I cannot think of any reason this would find a worse commit than the prior implementation. If there were some workflows where people where explicitly using merge commits and avoiding the rebase because they specifically didn't want the newer commit chosen then those workflows will be broken. I wouldn't expect such a workflow to exist, but I'm more confident on my knowledge of the git side than the flutter CI infrastructure side of this change.

**Pull request merged**